TYP: Groupby sum|prod|min|max|first|last methods

topper-123 opened the pull request:
This makes two things:
* makes mypy know the signatures of the agg methods (``sum``, ``prod``, etc.) on Groupby
* gives the agg methods on Groupby objects a return type

This ensues that mypy and  IDE's always know that the return type of e.g. ``df.groupby('a').sum()`` is a DataFrame, which is nice when chaining.

topper-123 committed:
CLN/TYP: Groupby agg methods

WillAyd commented on pandas/core/groupby/generic.py:
```diff
@@ -79,6 +79,28 @@
 if TYPE_CHECKING:
     from pandas.core.internals import Block
 
+_agg_template = """
```
@HH-MWB can the doc decorator simplify this?

topper-123 replied:
Ok, I've used the @doc decorator, saves a line for each method. The ``**kwargs`` in @doc should not be restricted to ``str``. So I've changed it to ``Any``.

@HH-MWB , would be nice to get some examples to @doc function, took a bit of reading before I understood it.

WillAyd commented on pandas/core/groupby/generic.py:
```diff
@@ -789,6 +811,36 @@ def count(self) -> Series:
 
+    @Substitution(f="sum", no=True, mc=0, return_type="Series")
+    @Appender(_agg_template)
+    def sum(self, numeric_only=True, min_count=0) -> Series:
```
Can you annotate `numeric_only` and `min_count` on these as well?

WillAyd reviewed (changes requested):
On board with the concept; a few things

topper-123 committed:
add parameter types + @doc decorator

WillAyd commented on pandas/util/_decorators.py:
```diff
@@ -245,7 +245,7 @@ def wrapper(*args, **kwargs) -> Callable[..., Any]:
 
 
-def doc(*args: Union[str, Callable], **kwargs: str) -> Callable[[F], F]:
+def doc(*args: Union[str, Callable], **kwargs: Any) -> Callable[[F], F]:
```
I think this should stay the way it is - can you revert changes to this module? Or is there a particular reason these wouldn't be a str?

topper-123 replied:
In this PR they’re bools and ints. Doesn’t make sense to restrict to str, IMO.

WillAyd replied:
Ah OK I see your point

WillAyd reviewed (approved):
lgtm

jbrockmendel commented:
there's got to be a less verbose way of doing this, isnt there?

jreback reviewed (changes requested):
why are you not generating these like we do in genetic for the stat functions ?

topper-123 commented:
> why are you not generating these like we do in genetic for the stat functions ?

Doing e.g. ``reveal_type(df.var)``on master gives a revealed type of ``Any``. I don't think it's possible to programatically add methods and also have mypy infer their types, parameters or return values. Mypy isn't advance enough to do that.

It's true that this is a bit repetitive, but always with slight variations. Getting correct parameter and return types is something I value highly in a library as complex as Pandas, so IMO this is worth it.

jbrockmendel commented:
> I don't think it's possible to programatically add methods and also have mypy infer their types, parameters or return values. Mypy isn't advance enough to do that.

cc @simonjayhawkins is this likely to improve at any point?  would something like stub files help?

I find this pattern annoying at debug-time, when it is an extra layer of calls to look through (since I have to double-check whether the super().foo(...) call is changing any kwargs).

(core.window.expanding is another place where we do something like this, but in that case for docstrings)

WillAyd commented:
FWIW I don't find stepping through the functions as is necessarily any clearer than inheriting as done here. This is more verbose, but I don't think factory-generating these really adds all that much value and I do think would be nice to support annotations on these

simonjayhawkins commented on pandas/core/groupby/generic.py:
```diff
@@ -1863,6 +1909,30 @@ def groupby_series(obj, col=None):
+    @doc(_agg_template, fname="last", no=False, mc=-1, return_type="DataFrame")
+    def last(self, numeric_only: bool = False, min_count: int = -1) -> DataFrame:
+        return super().last(numeric_only=numeric_only, min_count=min_count)
+
```
These are duplicates from above except the return values. Can you use Generic instead? The exception would be for unique docstrings for the Series/Frame methods. Does something need to be added to make these docstrings visible in the documentation.

topper-123 replied:
The doc string parameter ``return_type`` is also different, though. I could change these into

```python
@doc(_agg_template, fname="last", no=False, mc=-1, return_type="DataFrame")
def last(self, numeric_only: bool = False, min_count: int = -1) -> DataFrame:
        last_compat = partial(self._get_loc, loc=-1)

        return self._agg_general(
            numeric_only=numeric_only,
            min_count=min_count,
            alias="last",
            npfunc=last_compat,
        )
```

on both SeriesGroupby and DataFrameGroupby (setting 'Dataframe' and 'Series' the relevant places). That would be repetitipus also, though, but it would avoid setting methods on the NDFrame class...

simonjayhawkins replied:
sorry if I wasn't clear, in the docs I see https://pandas.io/docs/reference/api/pandas.core.groupby.GroupBy.min.html but nothing for DataFrameGroupBy.min, whereas for most other methods there is a DataFrameGroupBy.<method>

So my question is, how to make these visible, otherwise if the generic docstring is used, this duplication is unnecessary and GroupBy could be Generic to give the correct return type. see https://github.com/pandas-dev/pandas/pull/29480/files#r347723361 where the idea was rejected.

topper-123 replied:
Ok, I see your point now, it's the issue of the doc reference.

I don't have a good solution to that. I've added doc strings to GroupBy methods and copied them over to SeriesGroupby and DataFrameGroupby. SeriesGroupby and DataFrameGroupby need methods themselves in order to get the correct type hint. feels repetitive, but I don't think we can make it better without dropping the type hint.

topper-123 committed:
Add doc string to GroupBy methods

simonjayhawkins commented:
> there's got to be a less verbose way of doing this, isnt there?

@topper-123 commit making _Groupby a generic class. https://github.com/pandas-dev/pandas/commit/dccbd00a75f5f39c2979e1970635ce00d26c3a11

jreback commented on pandas/core/groupby/groupby.py:
```diff
@@ -1336,105 +1389,68 @@ def size(self):
+        )
+
+    @staticmethod
+    def _get_loc(x, axis: int = 0, *, loc: int):
```
what? 

a) don't use staticmethod
b) what the heck is this for

jreback replied:
I would be more ok with this if its not a nested function like this. I am also -1 on any use of staticmethod.

jreback commented on pandas/core/groupby/groupby.py:
```diff
@@ -1336,105 +1389,68 @@ def size(self):
-            numeric_only: bool = True,
-            min_count: int = -1,
-        ):
+        def get_loc_notna(x, loc: int):
```
what is this?

jreback reviewed (changes requested):
I would just add all of the functions here and avoid changing agg_general at all, where the changes you are attempting are non-trivial (do a pre-cursor PR).

ok conceptually with adding these methods to get better typing.

jreback reviewed (changes requested):
can you split this pr up

1) a PR to refactor groupby_function so its clear what is going on 
2) the typing of functions

WillAyd commented:
@topper-123 still active? I think this is a nice change and helpful when we think about bumping our mypy pin, so would love if there's an easy way to address comments

topper-123 commented:
I've been a bit busy lately. I can see if I can make something work over the coming days.

jreback commented:
can you merge master ping on green.

topper-123 commented:
I've updated the PR and removed the return types, so this is only refactoring ``_add_numeric_operations`` and make mypy discover the methods.

So, this does:

```python
>>> meth = df.groupby('x').sum
>>> reveal_type(meth)
Any  # master
Revealed type is 'def (numeric_only: builtins.bool =, min_count: builtins.int =) -> Any'  # this PR
```

The Travis failure looks unrelated.

topper-123 committed:
simplify first_compat and last_compat

topper-123 commented:
I've made a new commit that simplifies the PR wrt. the static method.

topper-123 commented:
Superceded by #34200.
